# Diabetes Information Portal
## Phase 1: Exploring the dataset
A learning project using Python and Jupyter notebook


In [3]:
import pandas as pd

In [4]:
import matplotlib.pyplot as plt

In [5]:
df = pd.read_csv('diabetes.csv')
print("Data loaded successfully")
print(f"Shape: {df.shape[0]} rows, {df.shape[1]} columns")

Data loaded successfully
Shape: 768 rows, 9 columns


In [4]:
df.head() # Shows first 5 rows as default

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

In [5]:
df.tail()  # Shows last 5 rows

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
763           10      101             76             48      180  32.9   
764            2      122             70             27        0  36.8   
765            5      121             72             23      112  26.2   
766            1      126             60              0        0  30.1   
767            1       93             70             31        0  30.4   

     DiabetesPedigreeFunction  Age  Outcome  
763                     0.171   63        0  
764                     0.340   27        0  
765                     0.245   30        0  
766                     0.349   47        1  
767                     0.315   23        0

In [6]:
df.info() # Gives a technical summary of DataFrame structure.

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               768 non-null    int64  
 1   Glucose                   768 non-null    int64  
 2   BloodPressure             768 non-null    int64  
 3   SkinThickness             768 non-null    int64  
 4   Insulin                   768 non-null    int64  
 5   BMI                       768 non-null    float64
 6   DiabetesPedigreeFunction  768 non-null    float64
 7   Age                       768 non-null    int64  
 8   Outcome                   768 non-null    int64  
dtypes: float64(2), int64(7)
memory usage: 54.1 KB


In [7]:
df.describe() #Provides statistical summaries for all numeric columns.

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   768.000000  768.000000     768.000000     768.000000  768.000000   
mean      3.845052  120.894531      69.105469      20.536458   79.799479   
std       3.369578   31.972618      19.355807      15.952218  115.244002   
min       0.000000    0.000000       0.000000       0.000000    0.000000   
25%       1.000000   99.000000      62.000000       0.000000    0.000000   
50%       3.000000  117.000000      72.000000      23.000000   30.500000   
75%       6.000000  140.250000      80.000000      32.000000  127.250000   
max      17.000000  199.000000     122.000000      99.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age     Outcome  
count  768.000000                768.000000  768.000000  768.000000  
mean    31.992578                  0.471876   33.240885    0.348958  
std      7.884160                  0.331329   11.760232    0.476951  
min      0.000000                  0.078000   21.000000    0.000000  
25%     27.300000                  0.243750   24.000000    0.000000  
50%     32.000000                  0.372500   29.000000    0.000000  
75%     36.600000                  0.626250   41.000000    1.000000  
max     67.100000                  2.420000   81.000000    1.000000

In [ ]:
print("Missing values per column.")
print(df.isnull().sum())

Method: df.isnull().sum()
Two methods chained together:
df.isnull()

What it does: Checks every cell in DataFrame and returns True if the cell is empty/missing, False if it has data
Returns: A DataFrame of the same size, but filled with True/False values

python# If this was ran just like this, we would get:
df.isnull()
# Output would be a huge table of True/False values for each cell

.sum()

What it does: Adds up the True values (True = 1, False = 0) for each column
Returns: A count of missing values per column


Power Automate Analogy: Think of this like a two-step flow:

Step 1 (isnull): Check each item in your data - "Is this empty?" → Mark Yes or No
Step 2 (sum): Count all the "Yes" marks for each column


Your Result: All columns show 0, meaning you have no missing data!

In [ ]:
print("Outcome distribution:")
print(df['Outcome'].value_counts())

Outcome distribution:
Outcome
0    500
1    268
Name: count, dtype: int64


Method: df['Outcome']
This is column selection - accessing a specific column from your DataFrame.
pythondf['Outcome']  # Gets the entire Outcome column

When you write df['Outcome'], you're saying "give me just the Outcome column from my table."

You can also think of it as:

df = your entire table
['Outcome'] = selecting just one column (like filtering to one field)

Method: .value_counts()

What it does: Counts how many times each unique value appears in that column
Returns: A summary showing each unique value and its count

pythondf['Outcome'].value_counts()
```

> **Business Logic Analogy:** This is like creating a summary report. If you had a "Status" field in your work order portal with values like "Open", "Closed", "Pending", running `.value_counts()` would tell you:
> - Open: 45
> - Closed: 132
> - Pending: 23

**Your Result Explained:**
```
Outcome
0    500   ← 500 patients do NOT have diabetes
1    268   ← 268 patients DO have diabetes
This tells you that your dataset is imbalanced - you have almost twice as many non-diabetic (0) cases as diabetic (1) cases.

Chaining Methods Together
Here's the powerful concept: you can chain methods together because each method returns a result that the next method can work with:
pythondf.isnull().sum()
#  ↓         ↓
#  Step 1    Step 2
pythondf['Outcome'].value_counts()
#  ↓           ↓
#  Get column  Count values

Pipeline Analogy: Like water flowing through pipes - the output of one method becomes the input for the next!


Quick Reference
Method          | Purpose                  | Example
.isnull()       | Check for missing values | df.isnull()
.sum()          | Add up values            | df.isnull().sum() 
df['column']    | Select a column          | df['Age']
.value_counts() | Count unique values      | df['Outcome'].value_counts()


In [9]:
# Count values in Age columns
df['Age'].value_counts()

Age
22    72
21    63
25    48
24    46
23    38
28    35
26    33
27    32
29    29
31    24
41    22
30    21
37    19
42    18
33    17
38    16
36    16
32    16
45    15
34    14
46    13
43    13
40    13
39    12
35    10
50     8
51     8
52     8
44     8
58     7
47     6
54     6
49     5
48     5
57     5
53     5
60     5
66     4
63     4
62     4
55     4
67     3
56     3
59     3
65     3
69     2
61     2
72     1
81     1
64     1
70     1
68     1
Name: count, dtype: int64

In [10]:
# Check if any row has missing data
df.isnull().any()

Pregnancies                 False
Glucose                     False
BloodPressure               False
SkinThickness               False
Insulin                     False
BMI                         False
DiabetesPedigreeFunction    False
Age                         False
Outcome                     False
dtype: bool

In [11]:
# Get multiple columns
df[['Age','BMI','Outcome']]

Age   BMI  Outcome
0     50  33.6        1
1     31  26.6        0
2     32  23.3        1
3     21  28.1        0
4     33  43.1        1
..   ...   ...      ...
763   63  32.9        0
764   27  36.8        0
765   30  26.2        0
766   47  30.1        1
767   23  30.4        0

[768 rows x 3 columns]

In [7]:
# Check which columns have suspicious zeros
columns_to_check = ['Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'BMI']
zero_data = {} 

for column in columns_to_check:
    zero_count = (df[column] == 0).sum()
    zero_percentage = (zero_count / len(df)) * 100
    zero_data[column] = zero_count # store the count
    print(f"{column}: {zero_count} zeros ({zero_percentage:.1f}%)")

Glucose: 5 zeros (0.7%)
BloodPressure: 35 zeros (4.6%)
SkinThickness: 227 zeros (29.6%)
Insulin: 374 zeros (48.7%)
BMI: 11 zeros (1.4%)


In [14]:
# convert dic into df to plot
zero_df = pd.DataFrame(list(zero_data.items()),columns=['Column','Zero_Count'])
print(zero_df)

          Column  Zero_Count
0        Glucose           5
1  BloodPressure          35
2  SkinThickness         227
3        Insulin         374
4            BMI          11


In [ ]:
# Now plot using the NEW dataframe you just created
zero_df.plot(kind='bar', x='Column', y='Zero_Count', 
             title='Number of Zeros in Each Column',
             legend=False)
plt.ylabel('Count of Zeros')
plt.xlabel('Column Name')
plt.xticks(rotation=45)  # Rotate labels so they don't overlap
plt.tight_layout()  # Make sure everything fits
plt.show()

In [8]:
# Replace zeros with NaN (Not a Number) to mark them as truly missing
import numpy as np

df_clean = df.copy()  # Make a copy so you don't lose original data

# Replace suspicious zeros with NaN
for column in columns_to_check:
    df_clean[column] = df_clean[column].replace(0, np.nan)

# Now you can see the real missing data count
print(df_clean.isnull().sum())

Pregnancies                   0
Glucose                       5
BloodPressure                35
SkinThickness               227
Insulin                     374
BMI                          11
DiabetesPedigreeFunction      0
Age                           0
Outcome                       0
dtype: int64


In [17]:
print(df_clean.head())

   Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6    148.0           72.0           35.0      NaN  33.6   
1            1     85.0           66.0           29.0      NaN  26.6   
2            8    183.0           64.0            NaN      NaN  23.3   
3            1     89.0           66.0           23.0     94.0  28.1   
4            0    137.0           40.0           35.0    168.0  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1  


In [18]:
print(df_clean.describe())

       Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   768.000000  763.000000     733.000000     541.000000  394.000000   
mean      3.845052  121.686763      72.405184      29.153420  155.548223   
std       3.369578   30.535641      12.382158      10.476982  118.775855   
min       0.000000   44.000000      24.000000       7.000000   14.000000   
25%       1.000000   99.000000      64.000000      22.000000   76.250000   
50%       3.000000  117.000000      72.000000      29.000000  125.000000   
75%       6.000000  141.000000      80.000000      36.000000  190.000000   
max      17.000000  199.000000     122.000000      99.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age     Outcome  
count  757.000000                768.000000  768.000000  768.000000  
mean    32.457464                  0.471876   33.240885    0.348958  
std      6.924988                  0.331329   11.760232    0.476951  
min     18.200000                  

Looking at the Columns we have 
Name        |  Type       | Description                        | Role
Outcome        Categorical   0(no diabetes)/1(diabetes)          Target
Pregnancies    Numerical     Number of times pregnant            Input
Glucose        Numerical     Plasma glucose concentration is 
                             an oral glucose tolerance test.     Input
BloodPressure  Numerical     Diastolic blood pressure (mmHg)     Input
SkinThickness  Numerical     Triceps skin fold thickness (mm)    Input
Insulin        Numerical     2-Hour serum insulin (mu U/ml)      Input
BMI            Numerical     Body mass index                     Input
DiabetesPedigree
Function       Numerical     Diabetes pedigree function          Input
Age            Numerical     Age (years)                         Input

Fields that may have zero = Outcome, Pregnancies

# Phase 2: Data Cleaning
## Overview
After exploring the dataset, we identified "fake zeros"...

## Note the double square brackets for more than one column

In [19]:
# Create a copy of the dataset
df_clean = df.copy()  # This preserves the original